## Import Libraries

In [2]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

## Import Data Files

In [3]:
# Creating Path
path = r'C:\Users\anush\2020 Instacart Basket Analysis'

In [4]:
# Importing orders_products_merged_derived.pkl File
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'orders_products_merged_derived.pkl'))

In [5]:
# Checking the Shape of the Dataframe
df_ords_prods_merge.shape

(32404859, 19)

##  Aggregated Mean Calculation for Entire Dataframe

In [6]:
# Grouping by 'department_id' and 'order_number', and Calculating Mean using agg() Function
df_ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## How do the results for the entire dataframe differ from those of the subset?
Comparing the subset and entire dataframe:
Overall Trends: The entire dataframe generally has higher mean order numbers across departments, indicating it captures a broader range of customer behaviours than the subset.
Notable Differences: Department 21 has the highest mean in both, but it’s higher in the subset (25.53) than the entire dataset (22.90). Department 10 greatly increases the dataframe as a whole (20.19 vs. 18.68).
Consistent Departments: Some departments, like 13 and 17, show little difference between the two datasets, while others, like 3, 9, and 12, have notable variations.
Overall, the entire dataframe reveals more frequent ordering patterns across departments compared to the subset.

## Create a Loyalty Flag

In [7]:
# Aggregating Data with transform() Function
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\anush\AppData\Local\Temp\ipykernel_18476\3473307984.py:2: FutureWarning: The provided callable <function max at 0x0000019112EFC7C0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [8]:
# Checking the Output
df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                      3.0        False                  5          0   both   
1                     20.0        False                  1          1   both   
2                      6.0        False                 20          0   both   
3                      NaN         True                 10          0   both   
4                      9.0        False                 11          1   both   
5                     26.0        False                  7          0   both   
6                     30.0        False                  2          0   both   
7                     30.0        False                  1          0   both   
8                     19.0        False                  1          0   both   
9                     15.0        False                  2          1   both   
10                     NaN         True                  1          0   both   
11                     7.0        False                  1          0   both   
12                     7.0        False                  3          1   both   
13                    12.0        False                  4          1   both   
14                     2.0        False                  8          0   both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly bu

In [9]:
# Creating a 'Loyalty' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# Examining Value Counts on 'loyalty_flag'
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [11]:
# Checking the Updated Output
df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    ...  first_order  add_to_cart_order  reordered  _merge    price_range_loc  \
0   ...        False                  5          0    both  Mid-range product   
1   ...        False                  1          1    both  Mid-range product   
2   ...        False                 20          0    both  Mid-range product   
3   ...         True                 10          0    both  Mid-range product   
4   ...        False                 11          1    both  Mid-range product   
5   ...        False                  7          0    both  Mid-range product   
6   ...        False                  2          0    both  Mid-range product   
7   ...        False                  1          0    both  Mid-range product   
8   ...        False                  1          0    both  Mid-range product   
9   ...        False                  2          1    both  Mid-range product   
10  ...         True                  1          0    both  Mid-range product   
11  ...        False                  1          0    both  Mid-range product   
12  ...        False                  3          1    both  Mid-range product   
13  ...        False                  4          1    both  Mid-range product   
14  ...        False                  8          0    both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy           Most orders        32   
1   Regularly busy  Regularly busy        Average ord

## Analyze Spending Habits by Loyalty Category

In [12]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customers tend to purchase lower-cost items, with an average price of $10.39, while new customers spend more, with a mean cost of $13.29. Regular customers fall in the middle, averaging $12.50 per item. All groups share a minimum price of $1.0, but the extremely high maximum cost of $99,999 suggests possible outliers or data errors.

## Create a Spending Flag

In [13]:
# Aggregating Data with transform() Function
df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\anush\AppData\Local\Temp\ipykernel_18476\1144848887.py:2: FutureWarning: The provided callable <function mean at 0x0000019112EFD1C0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [14]:
# Checking Relevant Columns 
df_ords_prods_merge[['user_id', 'avg_product_price']].head(15)

user_id  avg_product_price
0       138           6.935811
1       138           6.935811
2       709           7.930208
3       764           4.972414
4       764           4.972414
5       777           6.935398
6       825           5.957576
7       910           6.680000
8      1052           7.162500
9      1052           7.162500
10     1379           5.813636
11     1480           5.081633
12     1480           5.081633
13     1480           5.081633
14     1494           6.803797

In [15]:
# Creating a 'Spending' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [16]:
# Examining Value Count on 'spending_flag' Column
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [17]:
# Checking Releavent Columns
df_ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0       138           6.935811   Low spender
1       138           6.935811   Low spender
2       709           7.930208   Low spender
3       764           4.972414   Low spender
4       764           4.972414   Low spender
5       777           6.935398   Low spender
6       825           5.957576   Low spender
7       910           6.680000   Low spender
8      1052           7.162500   Low spender
9      1052           7.162500   Low spender
10     1379           5.813636   Low spender
11     1480           5.081633   Low spender
12     1480           5.081633   Low spender
13     1480           5.081633   Low spender
14     1494           6.803797   Low spender

## Create an Order Frequency Flag

In [18]:
# Aggregating Data with transform() Function
df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\anush\AppData\Local\Temp\ipykernel_18476\4152215344.py:2: FutureWarning: The provided callable <function median at 0x000001911301F600> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [19]:
# Checking Relevant Columns 
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0       138                     3.0                         8.0
1       138                    20.0                         8.0
2       709                     6.0                         8.0
3       764                     NaN                         9.0
4       764                     9.0                         9.0
5       777                    26.0                        11.0
6       825                    30.0                        20.0
7       910                    30.0                         6.0
8      1052                    19.0                        10.0
9      1052                    15.0                        10.0
10     1379                     NaN                        20.0
11     1480                     7.0                         7.0
12     1480                     7.0                         7.0
13     1480                    12.0                         7.0
14     1494                     2.0                         7.0

In [20]:
# Creating a 'Order Frequency' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_between_orders'] <= 20) & (df_ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [21]:
# Examining Value Counts on 'order_frequency_flag'
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [22]:
# Checking Relevant Columns of Updated Dataframe
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         8.0   
3       764                     NaN                         9.0   
4       764                     9.0                         9.0   
5       777                    26.0                        11.0   
6       825                    30.0                        20.0   
7       910                    30.0                         6.0   
8      1052                    19.0                        10.0   
9      1052                    15.0                        10.0   
10     1379                     NaN                        20.0   
11     1480                     7.0                         7.0   
12     1480                     7.0                         7.0   
13     1480                    12.0                         7.0   
14     1494                     2.0                         7.0   

   order_frequency_flag  
0     Frequent customer  
1     Frequent customer  
2     Frequent customer  
3     Frequent customer  
4     Frequent customer  
5      Regular customer  
6      Regular customer  
7     Frequent customer  
8     Frequent customer  
9     Frequent customer  
10     Regular customer  
11    Frequent customer  
12    Frequent customer  
13    Frequent customer  
14    Frequent customer

## Export the Dataframe in pickle Format to the "Prepared Data" Folder

In [23]:
# Exporting the Dataframe
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))